components/tools/OmeroJava/test/integration/ThumbnailStoreTest.java

```diff
@@ -229,4 +230,73 @@ public void testGetThumbnailAfterReset() throws Exception {
         }
         svc.close();
     }
+
+    /**
+     * Test that thumbnails can be retrieved from multiple groups at once.
+     * @throws Throwable unexpected
+     */
+    @Test
+    public void testGetThumbnailsMultipleGroups() throws Throwable {
+        final byte[] thumbnail;
+        final long pixelsIdα, pixelsIdβ;
+        ThumbnailStorePrx svc = null;
+
+        /* create a fake image file */
+        final File file = File.createTempFile(getClass().getSimpleName(), ".fake");
+        file.deleteOnExit();
+
+        try {
+            /* import the image as one user in one group and get its thumbnail */
+            pixelsIdα = importFile(importer, file, "fake", false).get(0).getId().getValue();
+            svc = factory.createThumbnailStore();
+            svc.setPixelsId(pixelsIdα);
+            thumbnail = svc.getThumbnailByLongestSide(null);
+        } finally {
+            if (svc != null) {
+                {
+                    svc.close();
+                    svc = null;
+                }
+            }
+        }
+
+        setUpNewUserWithImporter();
+
+        try {
+            /* import the image as another user in another group and get its thumbnail */
+            pixelsIdβ = importFile(importer, file, "fake", false).get(0).getId().getValue();
+            svc = factory.createThumbnailStore();
+            svc.setPixelsId(pixelsIdβ);
+            Assert.assertEquals(svc.getThumbnailByLongestSide(null), thumbnail);
+        } finally {
+            if (svc != null) {
+                {
+                    svc.close();
+                    svc = null;
+                }
+            }
+        }
+
+        final Map<Long, byte[]> thumbnails;
+
+        try {
+            /* use all-groups context to fetch both thumbnails at once */
+            final List<Long> pixelsIdsαβ = ImmutableList.of(pixelsIdα, pixelsIdβ);
+            final Map<String, String> allGroupsContext = ImmutableMap.of("omero.group", "-1");
+            svc = factory.createThumbnailStore();
+            thumbnails = svc.getThumbnailByLongestSideSet(null, pixelsIdsαβ, allGroupsContext);
```

Comment:
As from card,

> PR apparently allows thumbnails to be fetched in an all-groups context for only images that previously had thumbnail fetched otherwise.

the not-all-groups fetches appear to be needed for this step to work. Methods like `ThumbnailCtx.createMissingThumbnailMetadata` might be the key because calls like `updateService.saveAndReturnIds` do not work in a -1 context.